### PR TITLE
HPCC-13804 Minor improvements to HASH32/HASH64

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln please review

File: rtl/eclrtl/eclrtl.cpp [hunk 1]
```diff
@@ -3520,8 +3538,26 @@ hash64_t rtlHash64VUnicode(UChar const * k, hash64_t initval)
 unsigned rtlHash32Data(size32_t len, const void *buf, unsigned hval)
 {
     const unsigned char *bp = (const unsigned char *)buf;   /* start of buffer */
-    const unsigned char *be = bp + len;     /* beyond end of buffer */
 
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+    //This possibly breaks the aliasing rules for c++, but I can't see it causing any problems
+    while (len >= sizeof(unsigned))
```
Comment (richardkchapman) [inline rtl/eclrtl/eclrtl.cpp]: Assumes sizeof(unsigned) is 4, I think?

Comment (richardkchapman) [inline rtl/eclrtl/eclrtl.cpp]: (and why not do 64-bits?)

Comment (ghalliday) [inline rtl/eclrtl/eclrtl.cpp]: You're right it does - I should implement as a loop, and let it unroll.
I might time the difference for unsigned and unsigned8.  The bottleneck is the multiplies (on x86) which take 10 cycles each, so I suspect it won't speed it up by that much.

Comment (mckellyln) [inline rtl/eclrtl/eclrtl.cpp]: 8 helps a little bit over 4, probably only another 2%.
How large is len typically ?

Comment (ghalliday) [inline rtl/eclrtl/eclrtl.cpp]: It helps if it is a multiple of 8, but would slow it down if it wasn't.  This is used on string fields, so I'm not convinced...

**Comment by ghalliday** [comment]
@richardkchapman I have converted the code to use a loop.  I'm not sure about the unsigned __int64 change.  I can do it if you prefer.

**Comment by mckellyln** [comment]
Looks fine.  I would use __in64 if I new most of the time len >= 8.

Wish we could come up with a new hash that had separate sub-hashes that get multiplied independently and combined at the end.

**Pull request merged**